src/main/java/picard/analysis/directed/RnaSeqMetricsCollector.java

```diff
@@ -430,20 +430,29 @@ else if (this.metrics.SAMPLE != null) {
             return out;
         }
 
+        /**
+         * Derived class may override this method in order to return a different set of genes for
+         * picking transcripts for computing coverage, or to cache the value returned by geneOverlapDetector.getAll(),
+         * because it can be expensive to compute repeatedly.
+         */
+        protected Set<Gene> getGenesForPickTranscripts() {
+            return geneOverlapDetector.getAll();
+        }
+
         /** Picks the set of transcripts on which the coverage metrics are to be calculated. */
         public Map<Gene.Transcript, int[]> pickTranscripts(final Map<Gene.Transcript, int[]> transcriptCoverage) {
-            final Map<Gene.Transcript, Double> bestPerGene = new HashMap<Gene.Transcript, Double>();
+            final Set<Gene> genesWithTranscripts = getGenesWithTranscripts(transcriptCoverage.keySet());
 
             // Make a map of the best transcript per gene to it's mean coverage
-            for (final Gene gene : geneOverlapDetector.getAll()) {
+            final Map<Gene.Transcript, Double> bestPerGene = new HashMap<Gene.Transcript, Double>();
```

Comment:
You can use a diamond operator here, e.g. `new HashMap<>()`. the compiler infers the type parameters from the variable declaration.